src/s3path.jl

```diff
@@ -1,14 +1,26 @@
-struct S3Path <: AbstractPath
+struct S3Path{A<:AbstractAWSConfig} <: AbstractPath
     segments::Tuple{Vararg{String}}
     root::String
     drive::String
     isdirectory::Bool
-    config::AWSConfig
+    config::A
+
+    # this method ensures that there is proper conversion to Tuple{Vararg{String}} for first arg,
+    # which can be a little tempremental if you don't explicitly handle it
+    function S3Path{A}(segments, root::AbstractString, drive::AbstractString, isdirectory::Bool,
```

Comment:
Is there some more context to this to adding this in? Just not following how the change from `AWSConfig -> AbstractAWSConfig` necessitates this

Reply:
It's rather subtle.  It turns out that `FilePathsBase` calls type constructors directly somewhere.  For example, if you have a function
```julia
f(::Type{T}) where {T} = T()
```
if `T` is `S3Path{AWSConfig}`, this will fail unless the above method is defined, because this is a *different method* from `S3Path()` (because `S3Path{AWSConfig}` and `S3Path` are different types, the latter is an implicit abstract type).

I concede that the constructors I've wound up with here are a little messy, but I have included them because they should
- Ensure there are no breaking changes for the `S3Path` constructor, i.e. the user interface of AWSS3.jl has no breaking changes from this PR.
- Ensure that `FilePathsBase` has the methods it needs for the calls it makes.
- Ensure that the first argument of `S3Path` always gets converted to the correct type, which is a little subtle (due to it being a bit of a weird type) and seemed to be error prone before this PR.